File: lib/rbac.js
/**
 * Role-Based Access Control (RBAC) Matrix & Permission Middleware
 * Enforces least privilege across Admin, Analyst, Trader, Compliance, and Guest roles.
 */

export const ROLES = {
  ADMIN: 'admin',
  ANALYST: 'analyst',
  TRADER: 'trader',
  COMPLIANCE: 'compliance',
  GUEST: 'guest',
};

export const PERMISSIONS = {
  VAULT_READ: 'vault:read',
  VAULT_WRITE: 'vault:write',
  VAULT_DELETE: 'vault:delete',
  AUDIT_READ: 'audit:read',
  AUDIT_LOG: 'audit:log',
  TRADE_EXECUTE: 'trade:execute',
  POSITIONS_WRITE: 'positions:write',
  EXPORT_DATA: 'export:data',
  USER_MANAGE: 'user:manage',
  SETTINGS_WRITE: 'settings:write',
};

export const ROLE_PERMISSIONS = {
  [ROLES.ADMIN]: [
    PERMISSIONS.VAULT_READ,
    PERMISSIONS.VAULT_WRITE,
    PERMISSIONS.VAULT_DELETE,
    PERMISSIONS.AUDIT_READ,
    PERMISSIONS.AUDIT_LOG,
    PERMISSIONS.TRADE_EXECUTE,
    PERMISSIONS.POSITIONS_WRITE,
    PERMISSIONS.EXPORT_DATA,
    PERMISSIONS.USER_MANAGE,
    PERMISSIONS.SETTINGS_WRITE,
  ],
  [ROLES.ANALYST]: [
    PERMISSIONS.VAULT_READ,
    PERMISSIONS.AUDIT_READ,
    PERMISSIONS.EXPORT_DATA,
  ],
  [ROLES.TRADER]: [
    PERMISSIONS.VAULT_READ,
    PERMISSIONS.TRADE_EXECUTE,
    PERMISSIONS.POSITIONS_WRITE,
    PERMISSIONS.EXPORT_DATA,
  ],
  [ROLES.COMPLIANCE]: [
    PERMISSIONS.VAULT_READ,
    PERMISSIONS.AUDIT_READ,
    PERMISSIONS.AUDIT_LOG,
    PERMISSIONS.EXPORT_DATA,
  ],
  [ROLES.GUEST]: [
    PERMISSIONS.VAULT_READ,
  ],
};

export function hasPermission(role, permission) {
  if (!role || !permission) return false;
  const permissions = ROLE_PERMISSIONS[role.toLowerCase()] || [];
  return permissions.includes(permission);
}

export function createAuthMiddleware(requiredPermission) {
  return (req, res, next) => {
    const role = req.headers['x-user-role'] || (req.user && req.user.role) || ROLES.GUEST;
    if (!hasPermission(role, requiredPermission)) {
      return res.status(403).json({
        error: 'Forbidden: Insufficient permissions',
        requiredPermission,
        userRole: role,
      });
    }
    next();
  };
}

File: lib/rbac.py
"""
Role-Based Access Control (RBAC) Module for Python Daemon & Endpoints
"""

from enum import Enum
from typing import Set, Dict, Any
import datetime
import logging

class Role(str, Enum):
    ADMIN = "admin"
    ANALYST = "analyst"
    TRADER = "trader"
    COMPLIANCE = "compliance"
    GUEST = "guest"

class Permission(str, Enum):
    VAULT_READ = "vault:read"
    VAULT_WRITE = "vault:write"
    VAULT_DELETE = "vault:delete"
    AUDIT_READ = "audit:read"
    AUDIT_LOG = "audit:log"
    TRADE_EXECUTE = "trade:execute"
    POSITIONS_WRITE = "positions:write"
    EXPORT_DATA = "export:data"
    USER_MANAGE = "user:manage"
    SETTINGS_WRITE = "settings:write"

ROLE_PERMISSIONS: Dict[Role, Set[Permission]] = {
    Role.ADMIN: {
        Permission.VAULT_READ,
        Permission.VAULT_WRITE,
        Permission.VAULT_DELETE,
        Permission.AUDIT_READ,
        Permission.AUDIT_LOG,
        Permission.TRADE_EXECUTE,
        Permission.POSITIONS_WRITE,
        Permission.EXPORT_DATA,
        Permission.USER_MANAGE,
        Permission.SETTINGS_WRITE,
    },
    Role.ANALYST: {
        Permission.VAULT_READ,
        Permission.AUDIT_READ,
        Permission.EXPORT_DATA,
    },
    Role.TRADER: {
        Permission.VAULT_READ,
        Permission.TRADE_EXECUTE,
        Permission.POSITIONS_WRITE,
        Permission.EXPORT_DATA,
    },
    Role.COMPLIANCE: {
        Permission.VAULT_READ,
        Permission.AUDIT_READ,
        Permission.AUDIT_LOG,
        Permission.EXPORT_DATA,
    },
    Role.GUEST: {
        Permission.VAULT_READ,
    },
}

class AuthorizationManager:
    """Enforces RBAC and maintains structured access audits."""

    @staticmethod
    def has_permission(role: Role, permission: Permission) -> bool:
        return permission in ROLE_PERMISSIONS.get(role, set())

    @staticmethod
    def audit_access(user_id: str, action: str, resource: str, permitted: bool) -> None:
        logger = logging.getLogger("rbac_audit")
        logger.info(
            f"timestamp={datetime.datetime.now(datetime.timezone.utc).isoformat()} "
            f"user_id={user_id} action={action} resource={resource} permitted={permitted}"
        )
